### PR TITLE
[#92] Constructed identities

### LINKED ISSUE #92: Constructed identities
https://docs.ltonetwork.com/v/edge/protocol/identities/decentralized-identifiers-did#constructed-identities

* [x] Index associations with type `0x01??`. Store relevant information (not the whole tx) in an index specifically for DIDs.
* [x] Index association revocations. Update the indexed association, marking it as revoked.
* [x] When resolving a DID, lookup DID associations and add the verification methods.
* [ ] Improve the documentation with an example.

If the public key of the recipient of the association is unknown, ignore it. Don't add a verification method in that case.

### PR DESCRIPTION
**Pull request by brunodb3**
- Adds indexing method for association transactions DID verification
- Adds `verification method service` for constructing the DID methods out of the indexed information
- Changed the resolving of identities to include the verification methods information
- Adds revoking of the associations
- Adds a `npm test:watch` for easier testing locally

---

closes #92

### CODE REVIEW

File: src/identity/identity.service.ts [hunk 1]
```diff
@@ -35,33 +38,42 @@ export class IdentityService {
...
+        blockchainAccountId: `${id}@lto:${chainIdOf(id)}`,
+      }]
     };
+
+    for (const verificationMethod of verificationMethods) {
+      const didVerificationMethod = verificationMethod.asDidMethod(publicKey);
+      didDocument.verificationMethod.push(didVerificationMethod);
+      const didReference = [didVerificationMethod.id];
+
+      if (verificationMethod.isAuthentication()) didDocument.authentication = didReference;
```
Comment (jasny) [inline src/identity/identity.service.ts:70]: If there are multiple methods for authentication, the property will only contain the last one. You should do `didDocument.authentication.push(didVerificationMethod.id)`

File: src/identity/identity.service.ts [hunk 2]
```diff
@@ -35,33 +38,42 @@ export class IdentityService {
...
+      if (verificationMethod.isAssertionMethod()) didDocument.assertionMethod = didReference;
+      if (verificationMethod.isKeyAgreement()) didDocument.keyAgreement = didReference;
+      if (verificationMethod.isCapabilityInvocation()) didDocument.capabilityInvocation = didReference;
+      if (verificationMethod.isCapabilityDelegation()) didDocument.capabilityDelegation = didReference;
+    }
+
+    return didDocument;
```
Comment (jasny) [inline src/identity/identity.service.ts:77]: If there are **no** verification methods through associations, it's an implicit DID. In that case, `did:lto:${address}#key` needs  to be set as `authentication`, `assertionMethod`, and `capabilityInvocation`.

File: src/identity/identity.service.ts [hunk 3]
```diff
@@ -35,33 +38,42 @@ export class IdentityService {
...
+  async asDidDocument(id: string, publicKey: string): Promise<DIDDocument> {
+    const verificationMethods = await this.verificationMethodService.getMethodsFor(id);
+    const didDocument: DIDDocument = {
       '@context': 'https://www.w3.org/ns/did/v1',
-      'id': `did:lto:${id}`,
-      'verificationMethod': [{
-        id: `did:lto:${address}#key`,
+      id: `did:lto:${id}`,
+      verificationMethod: [{
+        id: `did:lto:${id}#key`,
```
Comment (jasny) [inline src/identity/identity.service.ts:57]: The `id` may contain `:derived:${secret}`, this should not be in the key of the verification method. It should just be the address. Therefore, pass the `address` as function param.

File: src/verification-method/model/verification-method.model.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import { chainIdOf } from "@lto-network/lto-crypto";
+
+export interface DIDVerificationMethod {
```
Comment (jasny) [inline src/verification-method/model/verification-method.model.ts:3]: Move this to the same file as the DIDDocument interface

File: src/verification-method/model/verification-method.model.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            id: `did:lto:${this.recipient}#key`,
+            type: 'Ed25519VerificationKey2018',
+            controller: `did:lto:${this.recipient}`,
+            publicKeyBase58: publicKey,
+            blockchainAccountId: `${this.recipient}@lto:${chainIdOf(this.recipient)}`,
+        }
+    }
+
+    public isAuthentication() {
+        const result = 0x0101 | this.relationships;
```
Comment (jasny) [inline src/verification-method/model/verification-method.model.ts:41]: Use contants for `0x101`, `0x102`, etc.

File: src/identity/identity.controller.spec.ts [hunk 6]
```diff
@@ -37,8 +39,9 @@ describe('DidController', () => {
     test('should get a identity document for the url', async () => {
```
Comment (jasny) [inline src/identity/identity.controller.spec.ts:39]: Test with
 * 0 additional verification methods (implicit DID)
 * 1 additional verification method
 * 2 (or more) additional verification methods

**Review by jasny — CHANGES_REQUESTED** [review]

File: src/identity/identity.service.ts (hunk 1 above)
Comment (brunodb3) [reply src/identity/identity.service.ts]: done

File: src/identity/identity.service.ts (hunk 2 above)
Comment (brunodb3) [reply src/identity/identity.service.ts:109]: done

File: src/identity/identity.service.ts (hunk 3 above)
Comment (brunodb3) [reply src/identity/identity.service.ts]: done

File: src/verification-method/model/verification-method.model.ts (hunk 4 above)
Comment (brunodb3) [reply src/verification-method/model/verification-method.model.ts]: done

File: src/verification-method/model/verification-method.model.ts (hunk 5 above)
Comment (brunodb3) [reply src/verification-method/model/verification-method.model.ts]: done

File: src/identity/identity.controller.spec.ts (hunk 6 above)
Comment (brunodb3) [reply src/identity/identity.controller.spec.ts:47]: done (created `identity.service.spec.ts` for proper unit testing)

File: src/identity/identity.service.spec.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+          blockchainAccountId: 'recipient@lto:chainId',
+        }, {
+          id: 'did:lto:secondRecipient#key',
+          type: 'Ed25519VerificationKey2018',
+          controller: 'did:lto:secondRecipient',
+          publicKeyBase58: 'publicKey',
+          blockchainAccountId: 'secondRecipient@lto:chainId',
+        }],
+        authentication: [
+          'did:lto:sender#key',
```
Comment (jasny) [inline src/identity/identity.service.spec.ts:160]: If there are one or more verification methods (through associations) for the identity, don't add `'did:lto:sender#key'` as `authentication`, `assertionMethod`, and `capabilityInvocation`.

File: src/verification-method/model/verification-method.model.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
+import { chainIdOf } from "@lto-network/lto-crypto";
+import { DIDVerificationMethod } from "identity/interfaces/identity.interface";
+
+interface MethodMap {
```
Comment (jasny) [inline src/verification-method/model/verification-method.model.ts:4]: This seems redundant / unnecessary

**Review by jasny — CHANGES_REQUESTED** [review]

File: src/verification-method/model/verification-method.model.ts (hunk 8 above)
Comment (brunodb3) [reply src/verification-method/model/verification-method.model.ts:4]: the interface `MethodMap` you mean?

File: src/identity/identity.service.spec.ts (hunk 7 above)
Comment (brunodb3) [reply src/identity/identity.service.spec.ts]: done

File: src/verification-method/model/verification-method.model.ts (hunk 8 above)
Comment (brunodb3) [reply src/verification-method/model/verification-method.model.ts]: done

**Review by jasny — APPROVED** [review]

**Pull request merged**